### ISSUE TITLE
若recompute的input包含模型input_tensor(stop_gradients=True)，会跳过重计算

### ISSUE THREAD

**Issue by johnyanccer** [PaddlePaddle/Paddle#52603]
### bug描述 Describe the Bug

前向重计算模块，如果传入recompute里的input是来自模型输入input，通常stop_gradients=True，则将warning：[Recompute]: None of the inputs to current recompute block need grad, therefore there is NO need to recompute this block in backward !

代码示例：https://www.paddlepaddle.org.cn/documentation/docs/zh/guides/06_distributed_training/data_parallel/recompute_cn.html#sishujubingxingxiadezhongjisuan

def get_fc_block(block_idx, input_size, is_last=False):
    block_name = "block_" + str(block_idx)
    block = paddle.nn.Sequential(
        (block_name + "_fc_0", paddle.nn.Linear(input_size, input_size, bias_attr=False)),
        (block_name + "_dropout", paddle.nn.Dropout(p=0.5)),
        (block_name + "_relu_1", paddle.nn.ReLU()),
        (block_name + "_fc_1", paddle.nn.Linear(input_size, input_size, bias_attr=False)),
        (block_name + "_relu_2", paddle.nn.ReLU()),
    )
    if is_last:
        block.add_sublayer(
            block_name + "_fc_2",
            paddle.nn.Linear(
                input_size, 1, bias_attr=False
            )
        )
    else:
        block.add_sublayer(
            block_name + "_fc_2",
            paddle.nn.Linear(input_size, input_size, bias_attr=False)
        )

    return block

class Naive_fc_net(paddle.nn.Layer):
    def __init__(self, input_size=10,
                recompute_blocks=[1, 3],
                recompute_kwargs={}):
        super(Naive_fc_net, self).__init__()
        self.recompute_blocks = recompute_blocks
        self.recompute_kwargs = recompute_kwargs
        self.runfunc0 = get_fc_block(0, input_size, is_last=False)
        self.runfunc1 = get_fc_block(1, input_size, is_last=False)
        self.runfunc2 = get_fc_block(2, input_size, is_last=False)
        self.runfunc3 = get_fc_block(3, input_size, is_last=False)
        self.runfunc4 = get_fc_block(4, input_size, is_last=True)
        self.total_func = [self.runfunc0, self.runfunc1, self.runfunc2, self.runfunc3, self.runfunc4]

    def forward(self, inputs):
        nums = len(self.total_func)
        for i in range(nums):
            if i in self.recompute_blocks:
                inputs = recompute(self.total_func[i], inputs)
            else:
                inputs = self.total_func[i](inputs)
        return inputs
def run_model(cuda_state, recompute_block=[], recompute_kwargs={}):
    gen = paddle.seed(10)
    gen.manual_seed(10)
    np.random.seed(10)
    random.seed(10)
    if cuda_state:
        paddle.set_cuda_rng_state(cuda_state)

    batch_size, input_size = 1, 10
    model = Naive_fc_net(
        input_size,
        recompute_blocks=recompute_block,
        recompute_kwargs=recompute_kwargs)
    optimizer = paddle.optimizer.SGD(learning_rate=0.01, parameters=model.parameters())
    loss_ = []
    param_ = []
    grad_ = []
    for _ in range(5):
        x_data = np.random.randn(batch_size, input_size).astype(np.float32)
        x = paddle.to_tensor(x_data)
        y_pred = model(x)
        loss = y_pred.mean()
        loss_.append(np.asarray(loss).tolist())
        loss.backward()
        optimizer.step()
        param_.append(np.asarray(model.parameters()[9]).tolist())
        grad_.append(np.asarray(model.parameters()[3]._grad_ivar()).tolist())
        optimizer.clear_grad()

    return loss_, param_, grad_

运行代码
loss, param, grad = run_model(cuda_state, recompute_block=[1, 3]) ，返回正常结果
loss, param, grad = run_model(cuda_state, recompute_block=[0, 3]) ，返回warning结果，因为recompute_block[0]的输入tensor来自模型输入，stop_gradients=True，这样就跳过了recompute_block[0]的重计算，结果也就不一样了

**Comment by LielinJiang** [comment]
recompute 本来就是在前向的时候不记录部分block的梯度所需的信息来节省显存的，如果input不需要梯度（stop_gradient=True）本身就会省略这些信息，所以就跳过重计算

**Comment by johnyanccer** [comment]
> recompute 本来就是在前向的时候不记录部分block的梯度所需的信息来节省显存的，如果input不需要梯度（stop_gradient=True）本身就会省略这些信息，所以就跳过重计算

可能我没描述清楚，假如我的input直接放入12层的transformer里面，按照recompute现有的逻辑，只能在transformer的encoder layer[1-11]层操作重计算，第[0]层没法操作重计算了，虽然这12层encoder layers结构完全一样，只有先后顺序之分，但是第一层就是没法操作重计算

**Comment by LielinJiang** [comment]
那就把input stop gradient设置成true咯，或者把第一层切的碎一点

**Comment by dongfangyixi** [comment]
不知道我理解的对不对。recompute 第二个输入是 checkpoint，如果是第0层，checkpoint 就是 原始数据输入，输入本来也不需要重计算的。

> > recompute 本来就是在前向的时候不记录部分block的梯度所需的信息来节省显存的，如果input不需要梯度（stop_gradient=True）本身就会省略这些信息，所以就跳过重计算
> 
> 可能我没描述清楚，假如我的input直接放入12层的transformer里面，按照recompute现有的逻辑，只能在transformer的encoder layer[1-11]层操作重计算，第[0]层没法操作重计算了，虽然这12层encoder layers结构完全一样，只有先后顺序之分，但是第一层就是没法操作重计算
